Add consensus and sync-related message types without nil struct fields

jordanschalm opened the pull request:
This PR implements a new approach to structural validity, by adding new inbound message types without nil struct fields. Replaces https://github.com/onflow/flow-go/pull/3470.

The primary purpose is to ensure safety against nil struct fields or slice entries. The secondary purpose is to begin to separate the internal (trusted) and external (untrusted) representations of a message, to enable eg. caching the ID of a trusted entity model.

This is still a draft - I will add documentation for the new types if we're happy with the general approach.

jordanschalm commented on model/messages/consensus.go:
```diff
@@ -4,11 +4,103 @@ import (
+	ExecutionDataID  flow.Identifier
+}
+
+func (ur UntrustedExecutionResult) ToFlowResult() *flow.ExecutionResult {
```
```suggestion
func (ur UntrustedExecutionResult) ToInternal() *flow.ExecutionResult {
```

jordanschalm commented:
In this PR I modified inbound message types, to use a distinct "untrusted" type for blocks and some block field types, in particular `Payload` and `ExecutionResult`, because these contained values which could be nil when received over the network.

So we have:
* `flow.Block` - the internal representation of the block
* `messages.UntrustedBlock` - the external representation of the block
  * contains duplicated `UntrustedPayload`, `UntrustedExecutionResult` types

These types are both nominally and structurally distinct, because of the difference in pointer fields and slice elements. They are also not consistently structurally distinct, because only sub-entities which contained potentially unsafe pointer fields were duplicated as a distinct `UntrustedXYZ` type. Other sub-entities are represented as the same type. This results in:
* verbose and unnecessary type conversion code
* less ability to share utility code which operates on core models

What we want instead, is types like:
* `flow.Block` - shared model, no notion of trusted vs untrusted
  * utility code and helper functions use this model
*  `messages.UntrustedBlock` - the external untrusted representation of the block
```go
// same core model structure, nominally distinct
type UntrustedBlock struct { 
  flow.Block
}
```
* `flow.TrustedBlock` - trusted, validated, internal representation of the block
```go
// same core model structure, nominally distinct, can embed additional computed data
type TrustedBlock struct {
  flow.Block
  computedID flow.Identifier
  // ... 
}
```

However, implementing this would require removing pointer fields and slice elements from the core types, which is a much bigger time investment than only adding the additional types. The fields that need to be changed are heavily used (>100 usages each) and individual usages can require more than trivial changes to update.

Given that end goal, this PR will introduce some technical debt in the form of duplicated types and conversion code.

durkmurder reviewed (commented):
It really depends on what path we want to take with our data structures. I think we can all agree that currently we use pointers too much, and our structures are not optimal. 
I see a few ways that we can take in long-run:

1. Complete update of our core data structure to use values instead of pointers.
(+) better concurrency guarantees 
(+) less stress on GC
(+) save time on calculating ID, ability to extend trusted data with pre-computed values
(+) no need to add extra logic for sanity checking wired messages.
(-) lots of refactoring 
(-) extra copy when converting from wire format

This approach means that we have a separate wire format `messages.UntrustredEntity` and convert it to internal trusted format `flow.TrustedEntity` which is built from transforming untrusted data. 

2. Use value based structure for wire messages and leave our core data structures as pointer based but with extra data.
(+) save time on calculating ID, ability to extend trusted data with pre-computed values
(+) no need to add extra logic for sanity checking wired messages.
(+) little effort
(-) extra conversion logic with extra copy when converting from wire format

This approach means that we use a separate wire format same as 1) but convert it exactly the way it's implemented in this PR. We won't introduce breaking changing to our core data structures but it still leaves us room for adding pre-computed ID or other data. 

3. Use the same wire format as we have right now but add runtime sanity checks for data validity + update core data structures to have more data.
(+) little effort
(+) save time on calculating ID, ability to extend trusted data with pre-computed values
(-) constant overhead for runtime checks
(-) more stress on GC
(-) we do checks which could be done automatically by support of type system.

This approach means that we won't change wire messages, we will still use pointers all over the place and rely on runtime checks to assert data validity. 

TL; DR; There is no clear winner, all of proposed solutions have their cons and pros, I think we need to have more discussions to select a path which we will take forward and stop collecting technical dept.

AlexHentschel commented on engine/collection/compliance/core.go:
```diff
@@ -81,17 +80,19 @@ func NewCore(
 // OnBlockProposal handles incoming block proposals.
 func (c *Core) OnBlockProposal(originID flow.Identifier, proposal *messages.ClusterBlockProposal) error {
-	header := proposal.Header
+	block := proposal.Block.ToInternal()
```
I am confused why this transformation `ToInternal()` is repeatedly happening while processing a block proposal: 
* https://github.com/onflow/flow-go/blob/4d1829104f468fdbe6a1399b457031006511a267/engine/collection/compliance/core.go#L199
* https://github.com/onflow/flow-go/blob/4d1829104f468fdbe6a1399b457031006511a267/engine/collection/compliance/core.go#L249

In my head, the conversion would be done once, which created certain guarantees in the resulting data structure. For example here, `ToInternal()` gives rise to the higher-level internal representation with certain guarantees, e.g. the relevant pointers are not nil. 

I would suggest to convert _once_ right here in the beginning of `OnBlockProposal`. (Probably beyond the scope of this PR, but that is where I would imagine the pre-computation of the ID would happen: also as part of the `ToInternal()` function.) If we want to work with `cluster.Block` from here on, we would need to update `PendingClusterBlockBuffer` to work with the internal representation instead, which I think would be worth it.

jordanschalm replied:
> In my head, the conversion would be done once, which created certain guarantees in the resulting data structure.

Good point. The multiple conversions (and usages) of the wire type is partly why I was unhappy with this approach - I think this will help to reduce the technical debt scope as well.

AlexHentschel commented on engine/common/follower/engine.go:
```diff
@@ -221,12 +219,12 @@ func (e *Engine) onBlockResponse(originID flow.Identifier, res *messages.BlockRe
 
 // onBlockProposal handles incoming block proposals. inRangeBlockResponse will determine whether or not we should wait in processBlockAndDescendants
 func (e *Engine) onBlockProposal(originID flow.Identifier, proposal *messages.BlockProposal, inRangeBlockResponse bool) error {
+	block := proposal.Block.ToInternal()
```
Also here, it would be more consistent in my mind, to convert once and then work with `flow.Block` from here on

AlexHentschel commented on engine/consensus/compliance/core.go:
```diff
@@ -93,21 +93,22 @@ func NewCore(
 
 // OnBlockProposal handles incoming block proposals.
 func (c *Core) OnBlockProposal(originID flow.Identifier, proposal *messages.BlockProposal, inBlockRangeResponse bool) error {
+	block := proposal.Block.ToInternal()
```
Would make the most sense to me, to only convert once here too (requiring changing the type of the pending block storage here as well)

AlexHentschel commented on model/messages/consensus.go:
```diff
@@ -4,11 +4,132 @@ import (
 type BlockProposal struct {
-	Header  *flow.Header
-	Payload *flow.Payload
+	Block UntrustedBlock
```
Just for curiosity, would there be a disadvantage to writing
```golang
type BlockProposal struct {
	UntrustedBlock
}
```
?

jordanschalm replied:
Not really, either seem OK to me.

AlexHentschel reviewed (approved):
Overall, I like the approach. It very nicely opens up the additional degree of freedom by decoupling the wire format from the internal representation. While I agree that it introduces some technical debt, I feel this PR is a good basis to allow further experimentation. I think blocks are a particularly good case for experimentation, because they (probably) have the most complex pipeline of verification.

jordanschalm commented:
* Wrote up our discussion from today [here](https://www.notion.so/dapperlabs/Message-validation-thoughts-and-proposal-ec9aee8ab6f949b8984cac5a42e235f1#ea70a7f6c4234ce083205b7db07f27b8)
* Updated this PR:
  * limited surface area of `messages.Block`-like wire types to one conversion per engine, removed from block buffer
  * removed `PendingBlock` types in favour of the `Slashable` interface Yurii introduced

durkmurder commented on module/buffer.go:
```diff
@@ -3,19 +3,18 @@ package module
 // children once the parent is received.
 type PendingBlockBuffer interface {
-	Add(originID flow.Identifier, proposal *messages.BlockProposal) bool
+	Add(originID flow.Identifier, block *flow.Block) bool
```
can we also accept `flow.Slashable` here?

durkmurder commented on module/buffer.go:
```diff
@@ -3,19 +3,18 @@ package module
 // cannot yet be processed because they do not connect to the rest of the chain
 // state. They are indexed by parent ID to enable processing all of a parent's
 // children once the parent is received.
 type PendingBlockBuffer interface {
```
We could also make this interface generic to work both with cluster blocks?

jordanschalm replied:
I have spent some time experimenting  with this - we can, but the exact approach has impacts beyond this module. I would prefer to spend some more time experimenting and choosing the right approach before merging anything. I've pushed all the WIP code to `jordan/3374--generics-experimentation-mess`.

durkmurder commented on module/buffer.go:
```diff
@@ -3,19 +3,18 @@ package module
+	ByID(blockID flow.Identifier) (*flow.Slashable[flow.Block], bool)
 
-	ByParentID(parentID flow.Identifier) ([]*flow.PendingBlock, bool)
+	ByParentID(parentID flow.Identifier) ([]*flow.Slashable[flow.Block], bool)
```
I would strongly prefer to use `flow.Slashable` instead of `*flow.Slashable` here since it's a structure that just holds two pointers, no need to allocate it on heap

jordanschalm replied:
Agree - good point

durkmurder commented on module/buffer.go:
```diff
@@ -28,11 +27,11 @@ type PendingBlockBuffer interface {
 // PendingClusterBlockBuffer is the same thing as PendingBlockBuffer, but for
 // collection node cluster consensus.
 type PendingClusterBlockBuffer interface {
```
Same here, I have a feeling we can get rid of this interface by using generics.

durkmurder commented on utils/unittest/fixtures.go:
```diff
@@ -198,28 +198,20 @@ func ProposalFixture() *messages.BlockProposal {
-		Header:   block.Header,
-		Payload:  block.Payload,
+// AsSlashable returns the input message T, wrapped as a flow.Slashable instance with a random origin ID.
+func AsSlashable[T any](msg *T) *flow.Slashable[T] {
```
Would strongly advocate for not using pointers for `flow.Slashable`

durkmurder commented on model/messages/collection.go:
```diff
@@ -24,12 +24,70 @@ type CollectionResponse struct {
+// cluster.Block type.
+// Deprecated: Please update cluster.Payload.Collection to use []flow.TransactionBody,
+// then replace instances of this type with cluster.Block
+type UntrustedClusterBlock struct {
```
Interested if we can use generics to deal with cluster blocks? Can we give it a try?

jordanschalm replied:
I did try this initially but didn't really find a simple way to apply generics here. It might be easier if `flow.Block` used generics too...

It isn't compiling but I'll throw it in `jordan/3374--generics-experimentation-mess` if you want to take a look at what I was trying. Maybe you have some different ideas for how to approach it.

durkmurder commented on model/events/synchronization.go:
```diff
@@ -1,16 +1,16 @@
 type SyncedBlock struct {
 	OriginID flow.Identifier
-	Block    *flow.Block
+	Block    messages.UntrustedBlock
```
Why are we using a value here? Have we considered using a pointer as before?

jordanschalm replied:
Because the structure is pretty small for copying, and we don't need reference semantics. I don't have a strong opinion though - do you think a pointer is better here?

durkmurder reviewed (approved):
Looks great, I really like the changes. Left some comments for discussion but nothing critical

jordanschalm commented on model/messages/consensus.go:
```diff
@@ -4,11 +4,132 @@ import (
+}
+
+// ToInternal returns the internal representation of the type.
+func (ur UntrustedExecutionResult) ToInternal() *flow.ExecutionResult {
```
https://github.com/onflow/flow-go/pull/3523/commits/0d4ecde53e46582fdbf081fe3b0e95ce8eeaab3a

jordanschalm commented:
bors merge
